components/support/guid/src/lib.rs

```diff
@@ -0,0 +1,482 @@
+            } else {
+                debug_assert!(v.iter().all(|b| b.is_ascii()));
+                Repr::Slow(unsafe { String::from_utf8_unchecked(v) })
+            })
+        })
+    }
+
+    /// Try to convert `s` into a guid. Equivalent to unwrapping
+    /// [`Guid::try_from_slice`]
+    ///
+    /// # Panics
+    ///
+    /// Panics if `s.len() >= MAX_GUID_LEN` or the bytes in `s` are not all
+    /// valid guid bytes.
+    #[inline]
+    pub fn from_slice(s: &[u8]) -> Self {
+        Guid::try_from_slice(s).unwrap()
+    }
+
+    /// Get the data backing this `Guid` as a `&[u8]`.
+    #[inline]
+    pub fn as_bytes(&self) -> &[u8] {
+        match &self.0 {
+            Repr::Fast(rep) => rep.bytes(),
+            Repr::Slow(rep) => rep.as_ref(),
+        }
+    }
+
+    /// Get the data backing this `Guid` as a `&str`.
+    #[inline]
+    pub fn as_str(&self) -> &str {
+        match &self.0 {
+            Repr::Fast(rep) => rep.as_str(),
+            Repr::Slow(rep) => rep.as_ref(),
+        }
+    }
+
+    /// Convert this `Guid` into a `String`, consuming it in the process.
+    #[inline]
+    pub fn into_string(self) -> String {
+        match self.0 {
+            Repr::Fast(rep) => rep.as_str().into(),
+            Repr::Slow(rep) => rep,
+        }
+    }
+
+    /// Returns true for Guids that are valid places guids, and false for all others.
+    pub fn is_valid_for_places(&self) -> bool {
+        self.len() == 12 && self.bytes().all(Guid::is_valid_places_byte)
+    }
+
+    /// Returns true if the byte `b` is a character that is allowed to
+    /// appear in a GUID.
+    #[inline]
+    pub fn is_valid_byte(b: u8) -> bool {
+        b' ' <= b && b <= b'~'
+    }
+
+    /// Returns true if the byte `b` is a valid base64url byte.
+    #[inline]
+    pub fn is_valid_places_byte(b: u8) -> bool {
+        BASE64URL_BYTES[b as usize] == 1
+    }
+}
+
+// This is used to implement the places tests.
+const BASE64URL_BYTES: [u8; 256] = [
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0,
+    0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1,
+    0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+];
+
+/// Represents a failure to create a Guid, due to it being not allowed
+/// on the server.
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub enum GuidError {
```

Comment:
Thinking more about this, I'm not sure GuidError is actually a good idea. Enforcing correctness would be a breaking change for anybody using completely garbage guids. 

How would you feel about making this infallible instead?

Reply:
I am not sure if this error is needed, because I don't know much about the Android/iOS code nor the server side implementation. But I have two naive questions that,
1. If we drop GuidError here, are we assuming the upstream caller should perform the check (on the `FfiStr` value)?
2. If we are not going to perform any validation on the guid value at all, will it cause server side error?

Reply:
> I am not sure if this error is needed

Yeah, it basically isn't. I was being a bit overzealous when I wrote it. We can lose the checks, or *maybe* log a warning if one happens.

>  1. If we drop GuidError here, are we assuming the upstream caller should perform the check (on the `FfiStr` value)?

No, we'd let them query with invalid guids. The querys will just return nothing. We never allow them to specify the guid when inserting data into the DB, so this is fine.

Specifically, The GuidErrors checks are actually more concerned with eventual use in `sync15`, where the server has restrictions on the kinds of guids which are allowed.`places` *does* have invariants in it's guids, but they're stricter than the sync servers -- and they are tested by `is_valid_for_places`.

> 2. If we are not going to perform any validation on the guid value at all, will it cause server side error?

In practice we generate all GUIDs we have locally, and are certain they are valid. We might want to add some to the guids we're about to upload when we integrate `SyncGuid` into `sync15`, but honestly this might be too much of an edge case to worry about, the server will handle this sort of error anyway.

Reply:
Thanks so much for answering my questions! I will try removing the GuidError within this PR then.